hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/blocking/BlockingResource.java

```diff
@@ -148,7 +148,17 @@ final void closeSession(long sessionId, List<Long> expiredWaitKeys, Map<Long, Ob
 
     final void collectAttachedSessions(Collection<Long> sessions) {
         sessions.addAll(getActivelyAttachedSessions());
-        for (WaitKeyContainer<W> key : waitKeys.values()) {
+        Collection<WaitKeyContainer<W>> values = waitKeys.values();
+        // We create an array with the copy of values to prevent the ConcurrentModificationException during
+        // iteration by values.
+        // We make the initial array size N-times bigger to reserve some additional space for values that
+        // could be added concurrently during copping values from LinkedHashMap
+        int arraySizeMultiplier = 2;
+        WaitKeyContainer[] waitKeyContainers = values.toArray(new WaitKeyContainer[values.size() * arraySizeMultiplier]);
```

Comment:
The values are `java.util.LinkedHashMap.LinkedValues`.
`toArray(T[] a)` on `LinkedValues` is implemented in `AbstractCollection`
In Java 8 it uses `iterator()`.

What am I missing?

Reply:
Thanks, @frant-hartm! Yes, it was my fault - I've used JDK15 to implement the fix. `toArray` method implementation has been changed at JDK 14: 
https://github.com/AdoptOpenJDK/openjdk-jdk14u/blob/4c6b476f2be142954d9823cc8a2e55a95d7433f4/src/java.base/share/classes/java/util/LinkedHashMap.java#L551-L557

Reply:
Depending on the implementation of toArray, this may or may not silence the exception but it won't fix the underlying problem, which is non-synchronized access to shared data. This can lead to undefined behavior.